tests/model/liverange.js

```diff
@@ -361,6 +362,147 @@ describe( 'LiveRange', () => {
 				expect( spy.calledOnce ).to.be.true;
 			} );
 		} );
+
+		describe( 'wrap', () => {
+			// NOTE: it overrides the variable defined globally in these tests.
+			// These tests need to be rewritten to use the batch API anyway and then this variable can be removed.
+			let live;
+
+			beforeEach( () => {
+				doc.schema.registerItem( 'p', '$block' );
+				doc.schema.registerItem( 'w' );
+
+				doc.schema.allow( { name: 'p', inside: 'w' } );
+				doc.schema.allow( { name: 'w', inside: '$root' } );
+			} );
+
+			afterEach( () => {
+				live.detach();
+			} );
+
+			it( 'is inside the wrapped range', () => {
+				setData( doc, '<p>x</p><p>[a]</p><p>x</p>' );
+
+				live = new LiveRange( doc.selection.getFirstPosition(), doc.selection.getLastPosition() );
+
+				// [<p>a</p>]
+				doc.batch().wrap( new Range( new Position( root, [ 1 ] ), new Position( root, [ 2 ] ) ), 'w' );
+
+				expect( stringify( root, live ) ).to.equal( '<p>x</p><w><p>[a]</p></w><p>x</p>' );
+			} );
+
+			it( 'its start is intersecting with the wrapped range', () => {
+				setData( doc, '<p>a[b</p><p>x</p><p>c]d</p>' );
+
+				live = new LiveRange( doc.selection.getFirstPosition(), doc.selection.getLastPosition() );
+
+				// [<p>ab</p>]
+				doc.batch().wrap( new Range( new Position( root, [ 0 ] ), new Position( root, [ 1 ] ) ), 'w' );
+
+				expect( stringify( root, live ) ).to.equal( '<w><p>a[b</p></w><p>x</p><p>c]d</p>' );
+			} );
+
+			it( 'its end is intersecting with the wrapped range', () => {
+				setData( doc, '<p>a[b</p><p>x</p><p>c]d</p>' );
+
+				live = new LiveRange( doc.selection.getFirstPosition(), doc.selection.getLastPosition() );
+
+				// [<p>cd</p>]
+				doc.batch().wrap( new Range( new Position( root, [ 2 ] ), new Position( root, [ 3 ] ) ), 'w' );
+
+				expect( stringify( root, live ) ).to.equal( '<p>a[b</p><p>x</p><w><p>c]d</p></w>' );
+			} );
+
+			it( 'its start is intersecting with the wrapped range (multilpe elements)', () => {
+				setData( doc, '<p>a[b</p><p>x</p><p>c]d</p>' );
+
+				live = new LiveRange( doc.selection.getFirstPosition(), doc.selection.getLastPosition() );
+
+				// [<p>ab</p><p>x</p>]
+				doc.batch().wrap( new Range( new Position( root, [ 0 ] ), new Position( root, [ 2 ] ) ), 'w' );
+
+				expect( stringify( root, live ) ).to.equal( '<w><p>a[b</p><p>x</p></w><p>c]d</p>' );
+			} );
+
+			it( 'its end is intersecting with the wrapped range (multiple elements)', () => {
+				setData( doc, '<p>a[b</p><p>x</p><p>c]d</p>' );
+
+				live = new LiveRange( doc.selection.getFirstPosition(), doc.selection.getLastPosition() );
+
+				// [<p>x</p><p>cd</p>]
+				doc.batch().wrap( new Range( new Position( root, [ 1 ] ), new Position( root, [ 3 ] ) ), 'w' );
+
+				expect( stringify( root, live ) ).to.equal( '<p>a[b</p><w><p>x</p><p>c]d</p></w>' );
+			} );
```

Comment:
Please add a test for `<p>x[x</p><p>foo</p><p>bar</p><p>y]y</p>` where `<p>foo</p><p>bar</p>` is wrapped.